src/state_circuit/state.rs

```diff
@@ -0,0 +1,1609 @@
+            let i = Expression::Constant(inv);
+
+            let q_stack_not_first = q_target_cur.clone()
+                * (q_target_cur.clone() - one.clone())
+                * (q_target_cur.clone() - two.clone())
+                * (four.clone() - q_target_cur.clone())
+                * i;
+
+            // q_stack_first is 12, we use 1/12 to normalize the value
+            let inv = F::from_u64(12_u64).invert().unwrap_or(F::zero());
+            let i = Expression::Constant(inv);
+
+            let q_stack_first = q_target_cur.clone()
+                * (two.clone() - q_target_cur.clone())
+                * (three - q_target_cur.clone())
+                * (four.clone() - q_target_cur)
+                * (q_target_next.clone() - one)
+                * (q_target_next.clone() - two)
+                * (four - q_target_next)
+                * i;
+
+            let address_cur = meta.query_advice(address, Rotation::cur());
+            let stack_address_table_zero =
+                meta.query_fixed(stack_address_table_zero, Rotation::cur());
+
+            let q_stack = q_stack_first + q_stack_not_first;
+
+            vec![(q_stack * address_cur, stack_address_table_zero)]
+        });
+
+        // global_counter is in the allowed range:
+        meta.lookup(|meta| {
+            let global_counter = meta.query_advice(global_counter, Rotation::cur());
+            let global_counter_table = meta.query_fixed(global_counter_table, Rotation::cur());
+
+            vec![(global_counter, global_counter_table)]
+        });
+
+        // Memory value (for non-first rows) is in the allowed range.
+        // Memory first row value doesn't need to be checked - it is checked above
+        // where memory init row value has to be 0.
+        meta.lookup(|meta| {
+            let q_target = meta.query_fixed(q_target, Rotation::cur());
+            let one = Expression::Constant(F::one());
+            let three = Expression::Constant(F::from_u64(3));
+            let four = Expression::Constant(F::from_u64(4));
+            let q_memory = q_target.clone()
+                * (q_target.clone() - one)
+                * (three - q_target.clone())
+                * (four - q_target);
+
+            let value = meta.query_advice(value, Rotation::cur());
+            let memory_value_table = meta.query_fixed(memory_value_table, Rotation::cur());
+
+            // q_memory is 4 when target is 2, we use 1/4 to normalize the value
+            let inv = F::from_u64(4_u64).invert().unwrap_or(F::zero());
+            let i = Expression::Constant(inv);
+
+            vec![(q_memory * i * value, memory_value_table)]
+        });
+
+        // padding is in the allowed range:
+        meta.lookup(|meta| {
+            let padding = meta.query_advice(padding, Rotation::cur());
+            let padding_value_table = meta.query_fixed(padding_value_table, Rotation::cur());
+
+            vec![(padding, padding_value_table)]
```

Comment:
Why not just use `padding * (1 - padding)` so that we can save a lookup table?

Reply:
Tried both versions, but using lookup seems to be much faster.

Reply:
I see. That's interesting. If that's true, does it apply to Read/Write flag column as well since we also check whether it's 0/1?

Reply:
@gaswhat Exactly, it applies to read/write flag too. Will do some more benchmarks on this though.

Reply:
Added a benchmark to measure the performance of checking the value being binary using a gate or a lookup. I was wrong, the gate approach is actually faster. I changed the padding constraint accordingly. Thanks for suggestion, @therealyingtong!